module/zfs/spa_errlog.c

```diff
@@ -136,83 +543,284 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		    NULL, NULL);
+
+		zep.zb_birth = bp.blk_birth;
+		rw_exit(&dn->dn_struct_rwlock);
+		dnode_rele(dn, FTAG);
+		dsl_dataset_rele(ds, FTAG);
+
+		if (error != 0 || BP_IS_HOLE(&bp))
+			continue;
+
+		uint64_t err_obj;
+		error = zap_lookup_int_key(spa->spa_meta_objset, *newobj,
+		    head_dataset_obj, &err_obj);
+
+		if (error == ENOENT) {
+			err_obj = zap_create(spa->spa_meta_objset,
+			    DMU_OT_ERROR_LOG, DMU_OT_NONE, 0, tx);
+
+			(void) zap_update_int_key(spa->spa_meta_objset,
+			    *newobj, head_dataset_obj, err_obj, tx);
+		}
+
+		char buf[64];
+		char *name = "";
+		errphys_to_name(&zep, buf, sizeof (buf));
+
+		(void) zap_update(spa->spa_meta_objset, err_obj,
+		    buf, 1, strlen(name) + 1, name, tx);
+	}
+	zap_cursor_fini(&zc);
+
+	VERIFY0(dmu_object_free(spa->spa_meta_objset, spa_err_obj, tx));
+}
+
+void
+spa_upgrade_errlog(spa_t *spa, dmu_tx_t *tx)
+{
+	uint64_t newobj = 0;
+
+	mutex_enter(&spa->spa_errlog_lock);
+	if (spa->spa_errlog_last != 0) {
+		sync_upgrade_errlog(spa, spa->spa_errlog_last, &newobj, tx);
+		spa->spa_errlog_last = newobj;
+	}
+
+	if (spa->spa_errlog_scrub != 0) {
+		sync_upgrade_errlog(spa, spa->spa_errlog_scrub, &newobj, tx);
+		spa->spa_errlog_scrub = newobj;
+	}
+	mutex_exit(&spa->spa_errlog_lock);
 }
 
 #ifdef _KERNEL
+/*
+ * If an error block is shared by two datasets it would be counted twice. For
+ * detailed message see spa_get_errlog_size() above.
+ */
 static int
-process_error_log(spa_t *spa, uint64_t obj, void *addr, size_t *count)
+process_error_log(spa_t *spa, uint64_t obj, void *addr, uint64_t *count)
 {
 	zap_cursor_t zc;
 	zap_attribute_t za;
-	zbookmark_phys_t zb;
 
 	if (obj == 0)
 		return (0);
 
-	for (zap_cursor_init(&zc, spa->spa_meta_objset, obj);
-	    zap_cursor_retrieve(&zc, &za) == 0;
-	    zap_cursor_advance(&zc)) {
+	if (!spa_feature_is_enabled(spa, SPA_FEATURE_HEAD_ERRLOG)) {
+		for (zap_cursor_init(&zc, spa->spa_meta_objset, obj);
+		    zap_cursor_retrieve(&zc, &za) == 0;
+		    zap_cursor_advance(&zc)) {
+			if (*count == 0) {
+				zap_cursor_fini(&zc);
+				return (SET_ERROR(ENOMEM));
+			}
+
+			zbookmark_phys_t zb;
+			name_to_bookmark(za.za_name, &zb);
+
+			if (copyout(&zb, (char *)addr +
+			    (*count - 1) * sizeof (zbookmark_phys_t),
+			    sizeof (zbookmark_phys_t)) != 0) {
+				zap_cursor_fini(&zc);
+				return (SET_ERROR(EFAULT));
+			}
+			*count -= 1;
 
-		if (*count == 0) {
-			zap_cursor_fini(&zc);
-			return (SET_ERROR(ENOMEM));
 		}
+		zap_cursor_fini(&zc);
+		return (0);
+	}
 
-		name_to_bookmark(za.za_name, &zb);
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, obj);
+	    zap_cursor_retrieve(&zc, &za) == 0;
+	    zap_cursor_advance(&zc)) {
 
-		if (copyout(&zb, (char *)addr +
-		    (*count - 1) * sizeof (zbookmark_phys_t),
-		    sizeof (zbookmark_phys_t)) != 0) {
-			zap_cursor_fini(&zc);
-			return (SET_ERROR(EFAULT));
+		zap_cursor_t head_ds_cursor;
+		zap_attribute_t head_ds_attr;
+		zbookmark_err_phys_t head_ds_block;
```

Comment:
I think the declaration of `head_ds_block` can be moved inside the loop below (just before it's passed to name_to_errphys()).